skbio/sequence/_substitution.py

```diff
@@ -0,0 +1,518 @@
+        >>> mat.alphabet
+        ('A', 'C', 'G', 'T')
+        >>> mat.scores
+        array([[ 1., -2., -2., -2.],
+               [-2.,  1., -2., -2.],
+               [-2., -2.,  1., -2.],
+               [-2., -2., -2.,  1.]])
+
+        """
+        alphabet = tuple(alphabet)
+        scores = np.identity(len(alphabet)) * (match - mismatch) + mismatch
+        return cls(alphabet, scores)
+
+    @classonlymethod
+    @experimental(as_of='0.5.10')
+    def by_name(cls, name):
+        """Load a pre-defined substitution matrix by its name.
+
+        Parameters
+        ----------
+        name : str
+            Name of the substitution matrix.
+
+        Returns
+        -------
+        SubstitutionMatrix
+            Named substitution matrix.
+
+        Raises
+        ------
+        ValueError
+            If named substitution matrix does not exist.
+
+        See Also
+        --------
+        get_names
+
+        Notes
+        -----
+        Available substitution matrix names can be obtained by ``get_names``.
+        Currently, the following names are supported:
+
+        - `NUC.4.4` (a.k.a. DNAfull): A nucleotide substitution matrix covering
+          all definite and degenerate nucleotides.
+
+        - Point Accepted Mutation (PAM) [1]_: A set of amino acid substitution
+          matrices, including `PAM30`, `PAM70` and `PAM250`.
+
+        - BLOcks SUbstitution Matrix (BLOSUM) [2]_: A set of amino acid
+          substitution matrices, including `BLOSUM45`, `BLOSUM50`, `BLOSUM62`,
+          `BLOSUM80` and `BLOSUM90`.
+
+        References
+        ----------
+        .. [1] Dayhoff, M., Schwartz, R., & Orcutt, B. (1978). A model of
+           evolutionary change in proteins. Atlas of protein sequence and
+           structure, 5, 345-352.
+        .. [2] Henikoff, S., & Henikoff, J. G. (1992). Amino acid substitution
+           matrices from protein blocks. Proceedings of the National Academy of
+           Sciences, 89(22), 10915-10919.
+
+        Examples
+        --------
+        >>> from skbio import SubstitutionMatrix
+        >>> mat = SubstitutionMatrix.by_name('BLOSUM62')
+        >>> len(mat.alphabet)
+        24
+        >>> mat['M', 'K']
+        -1.0
+
+        """
+        try:
+            return named_substitution_matrices[name]
+        except KeyError:
+            raise ValueError(f'Substitution matrix "{name}" does not exist.')
+
+    @classonlymethod
+    @experimental(as_of='0.5.10')
+    def get_names(cls):
+        """List names of pre-defined substitution matrices.
+
+        Returns
+        -------
+        list of str
+            Names of pre-defined substitution matrices.
+
+        See Also
+        --------
+        by_name
+        """
+        return list(named_substitution_matrices.keys())
+
+
+def _matrix_to_vector(mat):
+    """Convert a square matrix into a flattened vector consisting of the upper
+    triangle and the diagonal.
```

Comment:
```suggestion
    """Flatten a matrix to a vector of the upper triangle and diagonal.
```

This is very similar to condensed form availalbe in scipy?

Note the code does not assert the matrix is square

Reply:
Yes. However SciPy's `squareform` only supports matrices with zero diagonal. Therefore I had to write the code by hand.